### ISSUE TITLE
Should we experiment more?

### ISSUE THREAD

**Issue by appgurueu** [luanti-org/luanti#15548]
Because we guarantee backwards compatibility, there is excessive pressure to design APIs for new features "right" the *first time*, in the initial feature request or implementation, but because there usually is not a lot of real-world feedback at this stage, APIs may end up misguided.

But unfortunately much of the valuable feedback only pours in *after* the API becomes available in a stable release. Some of our modders pay attention to current engine development here on GitHub, but many don't (for multiple reasons).

I believe that the current process is too strict. We need to relax the backwards compatibility guarantee for some - typically larger - features: We should do more **experimental** APIs where we reserve ourselves the right to change them before they become stable. If something turns out bad in the first iteration, we can fix it without tedious deprecations and maintenance of legacy code paths.

An API being experimental could also include higher risk of bugs - we could leverage modders (and their users) as "out-in-the-wild" testers to an extent this way. This saves implementers from having to chase modders to test their implementations before they land: We could merge features with less testing scrutiny, knowing they are probably still not perfect and hoping to flesh it out after feedback comes in.

The motivation for this is that we will probably want to add some larger APIs, like v-rob's new UI API (#14263), or large CSM API changes, in the future.

**Comment by Zughy** [comment]
Side note as a server owner: we never use dev builds as we already have to deal with bugs and crashes we introduce from time to time when adding new features. Dev builds would mean having to deal with more issues AND not being sure that the error is our fault or not. Thus we stick to stable builds only, which is why we can't be betatesters

**Comment by cx384** [comment]
I agree, this would speed up development and approval of feature PRs, but I'm concerned that it will end up in a "There's nothing more permanent than a temporary solution" (translated German idiom) situation.
To prevent this, I suggest to make it strictly opt in, such that you have to manually enable it before you can use it in the API, e.g. `core.enable_experimental_feature("name")`

**Comment by sfan5** [comment]
This sounds good but I am concerned about ending up with "perpetually experimental" features, so there should be some rules to prevent that.
Something like: Within 2 releases of adding an experimental API it must either be marked stable, reworked or removed again.

**Comment by Zughy** [comment]
*looks at hypertext*

**Comment by nininik0** [comment]
This seems like a good idea, I'd love to get tons of new features even if experimental, would be fun.  The API seems to get better every update.

**Comment by v-rob** [comment]
The way I currently have things set up in #14263 is to hide all the new features behind a CMake flag `BUILD_UI`. The situation there is partly unique because the UI API requires `USE_SDL2`, so we can't enable the UI API in releases unconditionally anyway, even with a policy for experimental features in place.

> This sounds good but I am concerned about ending up with "perpetually experimental" features, so there should be some rules to prevent that.

This concerns me as well. I think it would be good to require that ContentDB mods don't use experimental features, both to mitigate this issue and to keep mods in ContentDB from breaking between releases.

**Comment by kromka-chleba** [comment]
Sounds good to me. I personally feel discouraged from testing PRs because there were many PRs that looked promising but failed to get merged after a long waiting time. On top of that compiling ~~Minetest~~ Luanti for every single new feature is tedious. The experimental status would give a little more confidence.

**Comment by rubenwardy** [comment]
It would suck if hypertext was suddenly removed as I use it in a few of my mods

However, if hypertext wasn't available in release builds to begin with then maybe it would have been finished and unmarked as experimental by now

**Comment by sfan5** [comment]
> In short, I don't know how many people would actually test these features if they might temporary and not accepted on CDB

Recognizing that something you've worked on is trash and better thrown away is a key part of experimentation.
For the whole idea to work modders need to be eager to take part in this experimentation by trying out new things knowing that they might not last.
If you can't justify this in your mods/games development then stick to the stable APIs.

(FWIW I disagree with the CDB suggestion)

**Comment by Zughy** [comment]
> For the whole idea to work modders need to be eager to take part in this experimentation by trying out new things knowing that they might not last

True, but practice tells us that it took two stable releases (~more than a year) to notice that the celestial vault API was quite messy, as apparently I was the only one who experimented with that and bothered enough to open several issues (and issues appeared also after that, e.g. fog #14176 ). What I mean is that, at the end of the day, I fear that basically no modders would actually try these features to provide useful feedback

**Comment by v-rob** [comment]
> Then almost no one would use them. Let's say I use an experimental feature in my mod, running on a public server.

Do note that running experimental features on servers is subject to some caveats. I have no intention of supporting network backwards compatibility in the UI API until it's out of beta, since I want to keep the possibility of major changes open according to modder feedback. That means that your server may only be able to support clients with the exact same version if you were to use the UI API. I don't know how much of an issue that sort of thing is for you, but it is something to keep in mind.

In regards to ContentDB: As a user, I prefer to only use mods in my worlds only if I think they have a pretty good chance at staying stable for years to come. I would prefer that there's at least some way to distinguish mods that use experimental features from those that don't. But then again, I suppose that there are mods under active development that break without warning, experimental features or no.

**Comment by appgurueu** [comment]
> But then again, I suppose that there are mods under active development that break without warning, experimental features or no.

I believe ContentDB has the "maintenance state" field for mods like this (which could be set to "beta" or "WIP"). It would seem sensible to require that packages which use experimental APIs set the maintenance state to one of those, which could then play more of a role in the in-game content browser.

**Comment by Xeno333** [comment]
I think this is a great idea though I think a seperate documentation should be created for all of the experimental features to set them apart from the stable ones.
